### PR TITLE
ng_at86rf2xx: fix wrong LQI and ensure 802.15.4 compliance with packet length field

### PR DESCRIPTION
**Pull request by daniel-k**
This PR should fix wrong LQI values for the new driver and adds additional measures to ensure IEEE 802.15.4 compliance regarding the PHR field as suggested by the datasheet and already discussed in #2825.

### CODE REVIEW

File: drivers/ng_at86rf2xx/ng_at86rf2xx.c [hunk 1]
```diff
@@ -233,9 +233,15 @@ void ng_at86rf2xx_tx_exec(ng_at86rf2xx_t *dev)
...
 size_t ng_at86rf2xx_rx_len(ng_at86rf2xx_t *dev)
 {
-    uint8_t res;
-    ng_at86rf2xx_fb_read(dev, &res, 1);
-    return (size_t)(res - 2);           /* extract the PHR and LQI field */
+    uint8_t phr;
+    ng_at86rf2xx_fb_read(dev, &phr, 1);
+
+    /* Mask MSB for 802.15.4 compliance, see
+     * section 37.1.1.2 of SAM R21 datahseet
```
Comment (haukepetersen) [inline drivers/ng_at86rf2xx/ng_at86rf2xx.c]: could you please reference the radio datasheets (-> it is section 8.1.1.2 for all four supported devices...)

File: drivers/ng_at86rf2xx/ng_at86rf2xx.c [hunk 2]
```diff
@@ -233,9 +233,15 @@ void ng_at86rf2xx_tx_exec(ng_at86rf2xx_t *dev)
...
+     */
+    phr &= ~(1 << 7);
+
+    return (size_t)(phr - 2);    /* subtract length of FCS field (checksum) */
```
Comment (haukepetersen) [inline drivers/ng_at86rf2xx/ng_at86rf2xx.c]: I would prefer putting this in a slightly more compact form
```c
/* ignore MSB (see datasheets section 8.1.1.2) and substract length of FCF field */
return (size_t)((phr - 2) & 0x7f);
```

File: drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c [hunk 3]
```diff
@@ -293,7 +293,8 @@ static void _receive_data(ng_at86rf2xx_t *dev)
     /* fill missing fields in netif header */
     netif = (ng_netif_hdr_t *)hdr->data;
     netif->if_pid = dev->mac_pid;
-    ng_at86rf2xx_rx_read(dev, &(netif->lqi), 1, pkt_len);
+    /* LQI offset: PHR (1 byte) + Payload (pkt_len) + FCS (2 byte) */
+    ng_at86rf2xx_rx_read(dev, &(netif->lqi), 1, pkt_len + 3);
```
Comment (haukepetersen) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c]: this seems still to be wrong: `ng_at86rf2xx_rx_read()` does not return the PHR. So shouldn't  `ng_at86rf2xx_rx_read(dev, &(netif->lqi), 1, pkt_len + 2)` be the correct solution?

Comment (daniel-k) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c]: Damn it, you're absolutely right! I misread the #if(n)def in `ng_at86rf2xx_rx_read()`, so it should actually be ` + 2`

File: drivers/ng_at86rf2xx/ng_at86rf2xx.c (hunk 2 above)
Comment (daniel-k) [inline drivers/ng_at86rf2xx/ng_at86rf2xx.c]: Done

**Comment by OlegHahm** [comment]
@haukepetersen, are we all set here and ready to go?

**Comment by daniel-k** [comment]
I would think so. Did anyone else test my changes?

**Comment by haukepetersen** [comment]
I test this as we speek

**Comment by haukepetersen** [comment]
Running this on the iot-lab and the samr21, I am not sure if the LQI values are correct (or actually I am not even sure what they should look like). If the LQI value should be fixed depending on the payload length, then everything works fine...

**Comment by jremmert-phytec-iot** [comment]
I tested the LQI value yesterday in conjunction with the kw2x-device and the driver of #2756. I received a value between 100 and 220 or so, depending on the rssi. When the distance between both devices was close, the value was around 220, when increasing the distance the value dropped. According to the reference manual of the kw2x-device (p.113) this behaviour is correct. The RSSI-value can be calculated from the LQI-value.

What should the RSSI-value look like in the NETIF-header? In general RSSI is the signal strength in dBm, so a value of -100 to -20 should be appropriate. Is the sign in this case just omitted?

**Comment by bapclenet** [comment]
On the samr21, I observe the same behaviour than @haukepetersen, LQI depends on the payload...

**Comment by haukepetersen** [comment]
So to me it still seems to broken with this PR?

**Comment by jremmert-phytec-iot** [comment]
If the at86rf2xx does not provide a working LQI, maybe you can calculate this value. In the Freescale KW2x device it is opposite, the RSSI-value is not provided; but it can be calculated using the LQI.
```RSSI = (LQI – 286.6) / 2.69333  [KW2x-Reference Manual]```

Do we really need both values for the upper layer? Since they seems to be directly related to each other one of the values would be sufficient.

**Comment by daniel-k** [comment]
> Do we really need both values for the upper layer? Since they seems to be directly related to each other one of the values would be sufficient.

Actually they aren't. RSSI and ED give you information about the signal **power** on your channel. This include of course noise and other signal sources (e.g. WLAN). LQI on the other hand is an indicator for the actual received signal **quality**.

> The radio transceiver uses correlation results of multiple symbols within a frame to determine the LQI value. This is done for each received frame.

---

> The received signal power as indicated by received signal strength indication (RSSI) value or energy detection (ED) value of the Atmel AT86RF233 do not characterize the signal quality and the ability to decode a signal.

---

> Various combinations of LQI and RSSI/ED are possible for routing decisions. As a rule of thumb, information on RSSI/ED is useful in order to differentiate between links with high LQI values. However, transmission links with low LQI values should be discarded for routing decisions, even if the RSSI/ED values are high, since it is merely an information about the received signal strength, whereas the source can be an interferer.

**Comment by haukepetersen** [comment]
> If the at86rf2xx does not provide a working LQI

Actually I think the devices do, it is just a question of reading the right value :-)

**Comment by daniel-k** [comment]
So I checked again and I can't seem to find the error. I looked at the raw framebuffer which looks good. Some header, payload, CRC and then 3 never changing bytes. I double-checked how Atmel is getting their LQI in lwMesh and they also read the first Byte after the CRC checksum as LQI.

I found out that LQI is only valid in 250kbp/s mode and when the payload is greater than 2 octets, which I checked is both met. 

I observed that even if the 3 Bytes after the checksum (FCS field) seem to be static when transmitting the same payload, the last 2 Bytes (ED, RX_STATUS) change when power cycling the board. That seems kind of strange, so I suspected that it's some kind of SPI or maybe data handling error.

That's where I am now. I had a look at the SPI driver. Even though there are some strange things happening, they should have no effect on my observations.

Maybe it's some kind of harfware fault? If nobody comes up with an idea I would open a ticket at Atmel support I guess.

**Comment by bapclenet** [comment]
@daniel-k, you were deep in your observations so oppening a ticket is what you should do now IMO.

**Comment by haukepetersen** [comment]
Asking Atmel about this seems the right way to go. @daniel-k, will you do that?

**Comment by daniel-k** [comment]
Already did 2 days ago, now waiting for a response. I also linked them this issue, so let's wait.

**Comment by bapclenet** [comment]
@daniel-k, any answer from ATMEL?

**Comment by daniel-k** [comment]
@bapclenet There was a reply, but no answer / solution yet. But we'll get there.

**Comment by bapclenet** [comment]
@daniel-k, any news then?

**Comment by daniel-k** [comment]
@bapclenet I was quite busy the last week, but today I managed to redo my measurements and sent them to Atmel again. They advised me to try their Wireless Performance Analyzer. I'll do that soon.

Here are my finding regarding LQI values: http://docdro.id/11n0h
The boards are lying 10cm away from each other and I'm sending packets from one to the other with increasing payload size. The numbers in the sheet are frame lengths, no payload lengths, that was a mistake.

This looks really strange, doesn't it?

**Comment by bapclenet** [comment]
>The LQI value is almost always 255 for scenarios with very low signal distortions and a signal strength much greater than the sensitivity level. In this case, the packet error rate tends towards zero and increase of the signal strength, that is by increasing the transmission power, cannot decrease the error rate any further. Received signal strength indication (RSSI) or energy detection (ED) can be used to evaluate the signal strength and the link margin.

So yes, this is definitely strange! At 10cm, it should be 255 everytime. Either the way we get LQI is wrong or there is an hardware problem. Let's see what ATMEL says about it.

**Comment by OlegHahm** [comment]
Maybe 10cm are too close and cause some kind of near field effect?

**Comment by daniel-k** [comment]
@OlegHahm I got similar results with incresed distance. And either way it would be a huge drawback in system design if this wouldn't work that way IMO.

**Comment by OlegHahm** [comment]
I think it was worth a try and it's not uncommon that radios have problems at low distances due to crosstalk.

**Comment by daniel-k** [comment]
I just went outside and arranged the boards at about 1m away from each other, turned Wifi of my notebook off. I got surprisingly low RSSI values, but the same (swinging) scheme of LQI.

**Comment by jremmert-phytec-iot** [comment]
> Maybe 10cm are too close and cause some kind of near field effect?

Some weeks ago I tested this setup with two KW2x radio modems. The closer the distance between both antennas was, the higher was the LQI I received. So I could not observe some kind of near field disturbion. But HF is always some kind of magic and generalization is difficult ;).

**Comment by daniel-k** [comment]
So Atmel wants me to flash their Performance Analyzer Software to test the board. Unfortunately I don't have access to a Windows PC at the moment and running Atmel Studio inside a VM doesn't seem to work for me (USB not recognized). Maybe someone else can try this?

They sent me these links:
 * https://www.youtube.com/watch?v=Rz5VxvnsEu0
 * http://www.atmel.com/webdoc/wirelesscomposer/index.html

**Comment by OlegHahm** [comment]
@kaspar030, haven't you found a way to run this Atmel software inside some Windows VM?

**Comment by bapclenet** [comment]
@daniel-k, @haukepetersen, ping

**Comment by daniel-k** [comment]
Nothing new from my side. Waiting for someone to try.

**Comment by bapclenet** [comment]
Ok, I'm on it, what do you me to do with the Performance Analyzer Software? Look at LQI value?

**Comment by daniel-k** [comment]
@bapclenet Thanks for your commitment!

I would have done their tests while watching the LQI value. See [this document](http://docdro.id/11n0h) where I ran tests with RIOT. From there it looks like LQI is coupled somehow to frame length OR there is something going wrong badly in our driver. But I don't know. And I never used Atmel Studio either ...

**Comment by bapclenet** [comment]
@daniel-k,
I change what they call frame lenght to 64 (I think it is different from your payload) and try to move it step by step up to 75. LQI value stays at 255 (or 254,253). It is definitely not what you've got which seems to be dependent of the payload.
Tell me if you want me to run other tests, look at the parameters I can change
http://www.atmel.com/webdoc/wirelesscomposer/wc_TxProIntroSec.html

**Comment by bapclenet** [comment]
capture

**Comment by daniel-k** [comment]
> My conclusion would be that we don't get the right LQI value, sorry.

Seems so. I really don't have a clue why. How are we going to proceed? If our investigation doesn't yield success, at least we should disable the LQI for this driver.

**Comment by bapclenet** [comment]
Have had a look at Atmel examples?

**Comment by bapclenet** [comment]
@daniel-k ping

File: drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c [hunk 4]
```diff
@@ -293,7 +293,8 @@ static void _receive_data(ng_at86rf2xx_t *dev)
     /* fill missing fields in netif header */
     netif = (ng_netif_hdr_t *)hdr->data;
     netif->if_pid = dev->mac_pid;
-    ng_at86rf2xx_rx_read(dev, &(netif->lqi), 1, pkt_len);
+    /* LQI offset: Payload (pkt_len) + FCS (2 byte) */
+    ng_at86rf2xx_rx_read(dev, &(netif->lqi), 1, pkt_len + 2);
```
Comment (thomaseichinger) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c]: `+ 2` doesn't hold for the rf231 as 

> If the SRAM read access is used to read an RX frame, the frame length field (PHR) cannot be accessed.

Compare section 9.3.2 User accessible Frame Content of the at86rf231 reference.
screen shot 2015-07-06 at 11 03 08

Comment (thomaseichinger) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c]: Looking at section 38.3.2 of the samr21 reference it seems there has to be one additional offset byte as PHR has offset 0. Right? 
screen shot 2015-07-06 at 11 10 40

While the offset should be fine for rf231 modules.

Comment (daniel-k) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c]: In general yes, but you need to have a look at the implementation:

```c
void ng_at86rf2xx_rx_read(ng_at86rf2xx_t *dev, uint8_t *data, size_t len,
                          size_t offset)
{
#ifndef MODULE_NG_AT86RF231
    ng_at86rf2xx_sram_read(dev, offset + 1, data, len);
#else
    ng_at86rf2xx_sram_read(dev, offset, data, len);
#endif
}
```

**Comment by daniel-k** [comment]
@bapclenet 
It seems like they don't use the LQI value, but normalize it or don't use it at all. There's a `#ifdef RSSI_TO_LQI_MAPPING` that decides which one to use.

In the case they use the LQI value, they only use the upper 2 bits and multiply this with the ED value (and something more, but constant).

I looked again in the datasheet of the RF233 with the hope to find something I missed, but nope. There are guys using the RF230 [1] that seem to get sensible LQI readings. So I more and more get the impression that something is wrong the RF chip inside the SAM-R21. At least I don't have a clue what should go wrong otherwise.

[1] http://www.avrfreaks.net/forum/lqi-reading-at86rf230

**Comment by OlegHahm** [comment]
If your assumption is right, we should be able to get proper readings on other at86rfxx boards, e.g. the IoT-LAB_M3, right?

**Comment by daniel-k** [comment]
See #3954 for fixing LQI. Masking PHR will have a separate PR.

**Pull request closed without merging**